src/util/expr_iterator.h

```diff
@@ -0,0 +1,228 @@
+/*******************************************************************\
+
+ Module: exprt iterator module
+
+ Author: DiffBlue Limited. All rights reserved.
+
+\*******************************************************************/
+
+#ifndef CPROVER_UTIL_EXPR_ITERATOR_H
+#define CPROVER_UTIL_EXPR_ITERATOR_H
+
+#include <iterator>
+#include <functional>
+#include <set>
+#include <algorithm>
+#include "expr.h"
+#include "preprocessor.h"
+
+// Forward depth-first search iterators
+// These iterators' copy operations are expensive, so use auto&, and avoid
+// std::next(), std::prev() etc.
+//
+// Incrementable only by calling ++it (to avoid accidental copying, as
+// this iterator is not cheap)
+//
+// Non-const iterators dereference to const exprt (for use with STL
+// algorithms) but have an extra .mutate() method. That method is used
+// to access non-const exprt reference but is an expensive operation
+
+// Naive depth-first-search iterator with .mutate method
+class depth_iteratort;
+// Naive depth-first-search iterator
+class const_depth_iteratort;
+// Depth first-search iterator with duplicate avoidance and a .mutate method
+// Slower iteration than naive version. Very expensive copy.
+class unique_depth_iteratort;
+// Depth first-search iterator with duplicate avoidance
+// Slower iteration than naive version. Very expensive copy.
+class const_unique_depth_iteratort;
+
+// Depth first search iterator - iterates over supplied expression
+// and all its operands recursively.
+// Base class using CRTP
```

Comment:
I might be wrong on this, but base class has to be adapted to work with CRTP, so in my book they all use CRTP.